File: package/bin/eventcount.c
#include <stdio.h>
#include <stdlib.h>
#include <sys/times.h>
#include <time.h>
#include <math.h>
#include <string.h>

#define BLOCK_SIZE 4096
#define TS_SIZE 50
#define INTERVAL 5

int main() {
    int         count;
    int         i;
    long        events;
    long        bytes;
    struct tms  timesout;
    long        ts;
    long        lastts;
    float       kbsec;
    float       kbsecavg;
    float       gbday;
    float       gbdayavg;
    char        *buf;
    char        *tsbuf;
    int         result;
    time_t      curtime;
    struct tm   *loctime;
    int         loopcount;

    lastts = times(&timesout);
    loopcount = 0;

    while (1) {
        buf = malloc(BLOCK_SIZE);
        tsbuf = malloc(TS_SIZE);

        result = fread(buf, 1, BLOCK_SIZE, stdin);
        if (result != BLOCK_SIZE) { 
            fputs ("Error reading from stdin", stderr); 
            exit (1);
        }

        count = 1;
        for (i=0;i<BLOCK_SIZE;i++) {
            if (buf[i] == '\n') {
                count++;
            }
        }
        // if (strncmp(&buf[BLOCK_SIZE-1], "\n", 1)) {
        //     events += count-1;  /* Unless the last character is a newline, we shouldn't count the last event
        //                             It will be picked up by the next iteration */
        // }
        events += count-1;
        bytes += BLOCK_SIZE;

        loopcount++;

        // Only check the time every 100 iterations
        if (loopcount % 100 == 0) {
            loopcount = 0;

            ts = time(NULL);
            // ts = (int)round(ts/1000);
            if (ts != lastts && ts % INTERVAL == 0 ) {
                // printf("TS: %ld\n", ts);
                lastts = ts;
                kbsec = bytes/INTERVAL/1024;
                gbday = kbsec * 60 * 60 * 24 /1024 / 1024;
                kbsecavg = (kbsecavg + kbsec) / 2;
                gbdayavg = (gbdayavg + gbday) / 2;

                loctime = localtime(&ts);

                strftime(tsbuf, TS_SIZE, "%Y-%m-%d %H:%M:%S", loctime);

                printf("%s Events/Sec: %1f Kilobytes/Sec: %1f GB/Day: %1f\n", tsbuf, (float)(events / INTERVAL), kbsec, gbday);

                bytes = 0;
                events = 0;
            }
        }

        free(buf);
        free(tsbuf);
    }
}

File: package/bin/eventcount.py
# Assumes events are one per line

from __future__ import division
from collections import deque
import datetime
import sys
import threading
import time


class Stats(threading.Thread):

	SLEEPTIME = 5
	AVERAGESAMPLESIZE = int( round((60*1)/SLEEPTIME, 0) ) # 1 Minutes of Data
	metrics = { 
		'count': 0,
		'bytes': 0,
		'kbsecavg': deque(maxlen=AVERAGESAMPLESIZE),
		'gbdayavg': deque(maxlen=AVERAGESAMPLESIZE)
	}

	def __init__(self):
		threading.Thread.__init__(self)

	def run(self):
		while True:
			kbsec = self.metrics['bytes']/self.SLEEPTIME/1024
			gbday = kbsec * 60 * 60 * 24 /1024 / 1024
			self.metrics['kbsecavg'].append( kbsec )
			self.metrics['gbdayavg'].append( gbday )
			kbsecavg = sum(list(self.metrics['kbsecavg'])) / len(self.metrics['kbsecavg'])
			gbdayavg = sum(list(self.metrics['gbdayavg'])) / len(self.metrics['gbdayavg'])
			print "%s Events/Sec: %s Kilobytes/Sec: %1f GB/Day: %1f" % \
						(datetime.datetime.strftime(datetime.datetime.now(), '%Y-%m-%d %H:%M:%S'),
						self.metrics['count'] / self.SLEEPTIME, kbsec, gbday )
			self.metrics['count'] = 0
			self.metrics['bytes'] = 0
			time.sleep(self.SLEEPTIME)




t = Stats()
t.daemon = True
t.start()


while True:
	# Bad Method
	# line = sys.stdin.readline()
	# t.metrics['count'] += 1
	# t.metrics['bytes'] += len(line)

	# Good Method
	stuff = sys.stdin.read(65536)
	t.metrics['count'] += stuff.count('\n')+1
	t.metrics['bytes'] += 65536